### PR TITLE
BUG: silent overflow in DateTimeArray subtraction

### LINKED ISSUE #22492: Incorrect timedelta computation of pd.Series datetime64[ns] if timedelta is too large
#### Problem description

Below I present two methods to calculate the time difference between two `pd.Timestamp`'s. The first way is the native python `datetime` subtraction, whereas the second way wraps things in a `pd.Series`. They should be equivalent. In example 1 the output is correct; but in example 2 the resulting timedelta is negative. This seems to be the case when the time difference  is greater than ~300 years. 

#### Code Sample
Example 1:
```python
t0 = pd.Timestamp('1723-08-23 00:00:00')
t1 = pd.Timestamp('2013-08-23 00:00:00')
print( 'Expected', (t1 - t0).total_seconds() )
print( 'Got\t', (pd.Series([t1]) - pd.Series([t0]))[0].total_seconds() )
```

Output 1:
```
Expected 9151574400.0
Got	 9151574400.0
```

Whereas Example 2:
```python
t0 = pd.Timestamp('1713-08-23 00:00:00')
t1 = pd.Timestamp('2013-08-23 00:00:00')
print( 'Expected', (t1 - t0).total_seconds() )
print( 'Got\t', (pd.Series([t1]) - pd.Series([t0]))[0].total_seconds() )
```
Output 2:
```
Expected 9467107200.0
Got	 -8979636873.709553
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.5.final.0
python-bits: 64
OS: Linux
OS-release: 4.4.0-130-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.23.4
pytest: None
pip: 18.0
setuptools: 40.2.0
Cython: None
numpy: 1.15.1
scipy: 1.1.0
pyarrow: None
xarray: None
IPython: 6.5.0
sphinx: None
patsy: None
dateutil: 2.7.2
pytz: 2018.5
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.3
openpyxl: None
xlrd: 1.1.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by shengpu-tang**
- [x] closes #22492
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by shengpu1126** [comment]
Hi @jbrockmendel , thanks for the follow-up. 
- With the suggested change, it now throws `OverflowError: Overflow in int64 addition`. Is this desired behavior? 
- Where should I add the tests? Maybe under `tests/arrays/test_datetimelike.py`?

**Comment by gfyoung** [comment]
@shengpu1126 : You should make the changes so that it gets the expected output that you reported in the original issue.  And yes, that does seem like a good place to put the test.  Don't forget a `whatsnew` entry too.

**Comment by jbrockmendel** [comment]
Test will go in tests.arithmetic.test_datetime64

File: doc/source/whatsnew/v0.24.0.txt [hunk 1]
```diff
@@ -582,6 +582,7 @@ Datetimelike
 - Bug in :class:`DataFrame` comparisons against ``Timestamp``-like objects failing to raise ``TypeError`` for inequality checks with mismatched types (:issue:`8932`,:issue:`22163`)
 - Bug in :class:`DataFrame` with mixed dtypes including ``datetime64[ns]`` incorrectly raising ``TypeError`` on equality comparisons (:issue:`13128`,:issue:`22163`)
 - Bug in :meth:`DataFrame.eq` comparison against ``NaT`` incorrectly returning ``True`` or ``NaN`` (:issue:`15697`,:issue:`22163`)
+- Bug in :class:`DatetimeIndex` subtraction that incorrectly failed to raise `OverflowError` (:issue:22492, :issue:22508)
```
Comment (jreback) [inline doc/source/whatsnew/v0.24.0.txt]: can you add a little more here, e.g. when this fails to raise

File: pandas/tests/arithmetic/test_datetime64.py [hunk 2]
```diff
@@ -1570,6 +1570,27 @@ def test_datetimeindex_sub_timestamp_overflow(self):
             with pytest.raises(OverflowError):
                 dtimin - variant
 
+    def test_datetimeindex_sub_datetimeindex_overflow(self):
+        dtimax = pd.to_datetime(['now', pd.Timestamp.max])
```
Comment (jreback) [inline pandas/tests/arithmetic/test_datetime64.py]: can you add a comment with the issue number

**Review by jreback — CHANGES_REQUESTED** [review]

**Review by jreback — CHANGES_REQUESTED** [review]
minor comments. ping on green (or at least except for 3.6/3.5 builds are currently failing to an unrelated error). so might want to wait a little bit to rebase.

File: doc/source/whatsnew/v0.24.0.txt [hunk 3]
```diff
@@ -582,6 +582,7 @@ Datetimelike
 - Bug in :class:`DataFrame` comparisons against ``Timestamp``-like objects failing to raise ``TypeError`` for inequality checks with mismatched types (:issue:`8932`,:issue:`22163`)
 - Bug in :class:`DataFrame` with mixed dtypes including ``datetime64[ns]`` incorrectly raising ``TypeError`` on equality comparisons (:issue:`13128`,:issue:`22163`)
 - Bug in :meth:`DataFrame.eq` comparison against ``NaT`` incorrectly returning ``True`` or ``NaN`` (:issue:`15697`,:issue:`22163`)
+- Bug in :class:`DatetimeIndex` subtraction that incorrectly failed to raise `OverflowError` when the difference between :class:`Timestamp`s (a :class:`Timedelta` object) exceeds `int64` limit (:issue:`22492`, :issue:`22508`)
```
[the remark this answers is not in the document]
Comment (shengpu1126) [reply doc/source/whatsnew/v0.24.0.txt]: sorry about that! 
for the record, it was mentioned [here](https://pandas.pydata.org/pandas-docs/stable/timedeltas.html#timedelta-limitations) that `pd.Timedelta`s are represented as nanoseconds using 64 bit integers, and I believe that's the source of potential overflow.

**Comment by jreback** [comment]
thanks!

**Pull request merged**